File: Leetcode/2061/2061.cpp
constexpr int STEPS = 4;
constexpr int MAXN = 305;

// 0: up 1: right 2: down 3: left
void turnRight(int& dir) {
    if (dir == 3) {
        dir = 0;
    }
    else {
        ++dir;
    }
}

void nextStep(int dir, int i, int j, int& new_i, int& new_j) {
    new_i = i;
    new_j = j;
    switch (dir) {
        case 0:
            --new_i;
            break;
        case 1:
            ++new_j;
            break;
        case 2:
            ++new_i;
            break;
        case 3:
            --new_j;
            break;
        default:
            break;
    }
}

bool vis[MAXN][MAXN][STEPS];

class Solution {
public:
    int numberOfCleanRooms(vector<vector<int>>& room) {
        memset(vis, 0, sizeof vis);
        int cur_i = 0, cur_j = 0, dir = 1;
        int new_i, new_j;
        int counter = 0;
        int n = (int)room.size();
        int m = (int)room[0].size();
        vis[cur_i][cur_j][dir] = true;
        ++counter;
        bool has_visited;
        while (true) {
            nextStep(dir, cur_i, cur_j, new_i, new_j);
            if (new_i >= 0 && new_i < n && new_j >= 0 && new_j < m && room[new_i][new_j] == 0) {
                if (vis[new_i][new_j][dir]) {
                    break;
                }
                vis[new_i][new_j][dir] = true;
                cur_i = new_i;
                cur_j = new_j;
                has_visited = false;
                for (int step = 0; step < STEPS; ++step) {
                    if (step != dir && vis[cur_i][cur_j][step]) {
                        has_visited = true;
                        break;
                    }
                }
                if (!has_visited) {
                    ++counter;
                }
            }
            else {
                turnRight(dir);
                if (vis[cur_i][cur_j][dir]) {
                    break;
                }
                vis[cur_i][cur_j][dir] = true;
            }
        }
        return counter;
    }
};

File: Leetcode/2061/2061.go
const STEPS = 4

func turnRight(dir *int) {
    if (*dir) == 3 {
		*dir = 0
	} else {
        (*dir)++
	}
}

func move(dir int, i int, j int, newI *int, newJ *int)  {
	switch dir {
	case 0:
		i--
	case 1:
		j++
	case 2:
		i++
	case 3:
		j--
	}
	*newI = i
	*newJ = j
}

func numberOfCleanRooms(room [][]int) int {
	n := len(room)
	m := len(room[0])
	vis := make([][][]bool, n)
	for i := 0; i < n; i++ {
		vis[i] = make([][]bool, m)
		for j := 0; j < m; j++ {
			vis[i][j] = make([]bool, STEPS)
		}
	}
	visStep := make([][]bool, n)
	for i := 0; i < n; i++ {
		visStep[i] = make([]bool, m)
	}
	curI, curJ, dir := 0, 0, 1
	var newI, newJ int
	counter := 1
	visStep[curI][curJ] = true
	vis[curI][curJ][dir] = true
	for {
		move(dir, curI, curJ, &newI, &newJ)
		if newI >= 0 && newI < n && newJ >= 0 && newJ < m && room[newI][newJ] == 0 {
			if vis[newI][newJ][dir] {
				break
			}
			vis[newI][newJ][dir] = true
			if !visStep[newI][newJ] {
				visStep[newI][newJ] = true
				counter++
			}
            curI, curJ = newI, newJ
		} else {
			turnRight(&dir)
			if vis[curI][curJ][dir] {
				break
			}
			vis[curI][curJ][dir] = true
		}
	}
	return counter
}